### ISSUE TITLE
[Research] How to enable library dependent instrumentations on .NET Framework

### ISSUE THREAD

**Issue by Kielek** [open-telemetry/opentelemetry-dotnet-instrumentation#968]
# Research

**Are you requesting automatic instrumentation for a framework or library? Please describe.**

For now, we have two library-dependent instrumentation in our code base:
* MongoDb,
* MySql.Data.

We expect more to implement in a similar way.
Both are working on .NET Core 3.1+.

I was able to execute `MySql.Data` instrumentation also on .NET Framework if the instrumented application was compiled with exactly the same version as the `OpenTelemetry.Instrumentation.MySqlData`.

**Potential solutions**

* Create a custom assembly resolver to resolve the dependency from the instrumented library. It could be tricky, especially if there is a non-standard resolver already used.
* Assembly redirects config file.
* Other ideas(?)

**Comment by pellared** [comment]
A custom assembly resolver could conflict with a user-defined assembly resolver.  It would have to be possible to disable this functionality. When it would be disabled (because the user has e a problem with it) then the customer could configure assembly redirects to resolve the assembly version conflict.

**Comment by pjanotti** [comment]
Assembly redirects seems to be the way to go in this case.

**Comment by nrcventura** [comment]
Some of the work with this issue involves investigating if there are other options, in addition to the binding redirects. For now we're simplifying things so that .net core and .net 6 users can try out the expanded support sooner.

On a separate note, I'm interested in understanding how many users of .net framework are interested in trying out this project.

**Comment by RassK** [comment]
Oh I have somehow skipped this issue.

I was thinking it could be done very similarly to MongoDB instrumentation.
`Type.GetType` will trigger assembly load if the middleman assembly is missing, otherwise it returns the user loaded type.
So all missed assembly loads will be redirected to loader and loader can pick it up from the main dir. 
.NET Core is using additional probing path and loads the newest version from the store, but since .NET Fx doesn't have additional probing paths configurable like that, the dll must exist in the OTel Autoinstrumentation dir or in the application dir.

If we add new library project (similar like `OpenTelemetry.AutoInstrumentation.AdditionalDeps`) but for .NET Fx only and reference all of the middleman nuget packages there, then during the build they will be placed exactly to the right dir.

Conflict resolution is still the same:
* .NET Core needs adjusting nuget package
* .NET Fx needs binding redirects

NB: Conflict in this case means user is referencing the middleman assembly but we don't like the referenced version.

**Comment by pjanotti** [comment]
The most ambitious and flexible solution would be to re-write the IL tokens at execution time and completely hide the issue from the user. The re-writing would be on the bytecode IL: it would be compiled against specific versions but would be re-written as the versions that are available at runtime. This needs new code for the IL rewrite used by bytecode instrumentation but it is the way to do it transparently. It can also make easier to write the bytecode instrumentations: they can be written in a fashion that actually uses the actual target types and methods making it easier to create and more accessible to the developer community in general.

It is unclear to me the potential performance impact of this compared with the current approach.

At this point I'm unsure if it is worth the trouble: are there enough bytecode instrumentations to justify the effort?

**Comment by pjanotti** [comment]
ILMerge, see #1418, may also be considered.

**Comment by pjanotti** [comment]
A combination of #1825 plus #1824 should suffice here. So after #1824 is merged we can revisit this item.

cc @RassK

**Comment by Kielek** [comment]
Lets check if MongoDB can work after @RassK  changes. If not, lets postpone after 1.0.0.

It is unlikely that MySql.Data will work for all cases with current support (changes in public API even between 8.0.30 and 8.0.31).

**Comment by Kielek** [comment]
Consider closing this issue. MongoDB is working fine both on .NET Framework and .NET.
Other libraries probably can be instrumented in the similar way.

**Comment by pjanotti** [comment]
We have automatic redirection and a kind of "roll forward" for .NET Framework, plus, the model to instrument libraries on .NET Framework. Closing per discussion above.
